lib/addons/src/types.ts

```diff
@@ -93,8 +94,13 @@ export interface OptionsParameter extends Object {
 
 export type StoryGetter = (context: StoryContext) => any;
 
-export type LegacyStoryFn<ReturnType = unknown> = (p?: StoryContext) => ReturnType;
-export type ArgsStoryFn<ReturnType = unknown> = (a?: Args, p?: StoryContext) => ReturnType;
+// This is the type of story function passed to a decorator -- does not rely on being passed any context
+export type PartialStoryFn<ReturnType = unknown> = (p?: StoryContextUpdate) => ReturnType;
+// This is a passArgsFirst: false user story function
+export type LegacyStoryFn<ReturnType = unknown> = (p: StoryContext) => ReturnType;
+// This is a passArgsFirst: true user story function
+export type ArgsStoryFn<ReturnType = unknown> = (a: Args, p: StoryContext) => ReturnType;
```

Comment:
I tightened some of these types up, by introducing a `PartialStoryFn` type (the thing a decorator expects in its first argument). 

I'm not sure if that's going to cause other problems but it seems like "A good thing".

Reply:
LGTM

Reply:
I am currently having to update a *lot* of things to get bootstrap to pass. I wonder if I should split it into a different PR.

Reply:
Yeah, we should prioritize making the runtime work correctly first, considering the code is very dynamic and complex, meaning the types are too.

Also this code is used in many places. So making the types strict is going to be a lot of work.

Deferring the "stricter types" work sounds reasonable to me.